File: LeetCode/42. Trapping Rain Water/solution.java
class Solution {
  public int trap(int[] height) {
      int left = 0;
      int right = height.length - 1;
      int maxHeightLeft = height[left];
      int maxHeightRight = height[right];
      int water = 0;
      while (left < right) {
          if (maxHeightLeft < maxHeightRight) {
              left++;
              if (height[left] < maxHeightLeft) {
                  water += maxHeightLeft - height[left];
              } else {
                  maxHeightLeft = height[left];
              }
          } else {
              right--;
              if (height[right] < maxHeightRight) {
                  water += maxHeightRight - height[right];
              } else {
                  maxHeightRight = height[right];
              }
          }
      }
      return water;
  }
}

File: LeetCode/42. Trapping Rain Water/solution.py
class Solution:
    def trap(self, height: List[int]) -> int:
        left = 0
        right = len(height) - 1
        max_height_left = height[left]
        max_height_right = height[right]
        water = 0
        while left < right:
            if max_height_left < max_height_right:
                left += 1
                if height[left] < max_height_left:
                    water += max_height_left - height[left]
                else:
                    max_height_left = height[left]
            else:
                right -= 1
                if height[right] < max_height_right:
                    water += max_height_right - height[right]
                else:
                    max_height_right = height[right]
        return water